Seccompiler documentation and design (#3)
rust-vmm/seccompiler: README.md

# Seccompiler

> Provides easy-to-use linux [seccomp-bpf](https://www.kernel.org/doc/html/latest/userspace-api/seccomp_filter.html) jailing.

Reviewer: ```suggestion
Provides easy-to-use Linux [seccomp-bpf](https://www.kernel.org/doc/html/latest/userspace-api/seccomp_filter.html) jailing.
```

> Seccomp is a safety net that can be applied on a process, to restrict a

Author: Sounds better, thanks

malicious user, that found a way of running arbitrary code, from executing
potentially dangerous system calls.

Writing BPF programs by hand is difficult and error-prone. This crate provides
a library of high-level wrappers over system call filtering, as well as
functions for generating and installing the filters.

## Supported platforms

Due to the fact that seccomp is a linux-specific feature, this crate is

> supported only on linux systems.

Reviewer: ```suggestion
supported only on Linux systems.
```

> Supported host arhitectures:

Reviewer: ```suggestion
Supported host architectures:
```

> - Little-endian x86_64

Author: A couple of things.

In regards to architecture, the system call numbers are different. When expressing JSON filters, the user specifies the syscall name, that is translated by seccompiler under the hood, to the arch-specific number.
Even if the user is not using the JSON functionality, seccomp-bpf filters are recommended to embed an architecture validation step at the beginning, to prevent the filter from being installed on a different arch (which will inherently have different syscall numbers).

And in regards to Big Endian, the BPF filter would be different. For example, if you want to compare an argument against a 64-bit integer, you need to split the integer into most significant and least significant halves, because the BPF has a 32-bit auxiliary register.

> - Little-endian x86_64

Author: There are definitely ways of making the crate available on multiple architectures, but I think we should do this on a per-need basis, since it has a dev and maintenance cost

> - Little-endian x86_64

Author: you mean having a `#![cfg(target_endian = "little")]` in the code? If so, will definitely do that in a future PR.

> - Little-endian x86_64

Reviewer 2: Yap.

> - Little-endian aarch64.

Reviewer: ```suggestion
- Little-endian aarch64
```

## Short seccomp tutorial

Linux supports seccomp filters as BPF programs, that are interpreted by the
kernel before each system call.

They are installed in the kernel using 

> [`prctl(PR_SET_SECCOMP)`](https://man7.org/linux/man-pages/man2/prctl.2.html)

Author: They cannot be invalidated per se. A process can have multiple filters and once a filter is installed on a process, it cannot be uninstalled.
If there are multiple filters for a thread, they are evaluated in the order they were installed and the most restrictive action is taken. i.e. If the first filter kills the process on `exec` but the second filter allows it, the process will be killed.

There are actually some best practices for seccomp (like having an allowlist vs a denylist) but they're mentioned in the man pages for `prctl`/`seccomp` and here https://www.kernel.org/doc/html/latest/userspace-api/seccomp_filter.html. Do you think it's worth having a similar section?

> [`prctl(PR_SET_SECCOMP)`](https://man7.org/linux/man-pages/man2/prctl.2.html)

Author: my bad, I was assuming that the first filter denied the `exec` while the second allowed it, in which case it cannot be invalidated.
But if it's the other way around, it can be further restricted.
I'll think about adding a `Recommendations` section

> [`prctl(PR_SET_SECCOMP)`](https://man7.org/linux/man-pages/man2/prctl.2.html)

Reviewer 2: In the recommendation section you can also link the other recommendations that you found.

> [`prctl(PR_SET_SECCOMP)`](https://man7.org/linux/man-pages/man2/prctl.2.html)

Author: yes, I added a `Seccomp best practices` section with all recommendations I know of.

or
[`seccomp(SECCOMP_SET_MODE_FILTER)`](https://man7.org/linux/man-pages/man2/seccomp.2.html).

As input, the filter receives a C struct of the following type:

```C
struct seccomp_data {
    int nr; // syscall number.
    __u32 arch; // arch-specific value for validation purposes.
    __u64 instruction_pointer; // as the name suggests..
    __u64 args[6]; // syscall arguments.
};
```

In response, a filter returns an action, that can be either one of:

```C
#define SECCOMP_RET_KILL_PROCESS 0x80000000U /* kill the process */
#define SECCOMP_RET_KILL_THREAD  0x00000000U /* kill the thread */
#define SECCOMP_RET_TRAP         0x00030000U /* disallow and force a SIGSYS */
#define SECCOMP_RET_ERRNO        0x00050000U /* returns an errno */
#define SECCOMP_RET_TRACE        0x7ff00000U /* pass to a tracer or disallow */
#define SECCOMP_RET_LOG          0x7ffc0000U /* allow after logging */
#define SECCOMP_RET_ALLOW        0x7fff0000U /* allow */
```

## Design

In seccompiler, filters are defined either with a JSON file or using Rust code,
with library-defined structures. Both representations are semantically
equivalent and model the rules of the filter. Choosing one or the other depends

> on the use case and preference.

Reviewer: Cool :+1:

The core of the package is the module responsible for the BPF compilation.
It compiles seccomp filters expressed as Rust code, into BPF filters, ready to

> be loaded into the kernel. This is the seccompiler **backend.**

Reviewer: ```suggestion
be loaded into the kernel. This is the seccompiler **backend**.
```

The process of translating JSON filters into BPF goes through an extra step of
deserialization and validation (the **JSON frontend**), before reaching the
same backend for BPF codegen.

The Rust representation is therefore also an Intermediate Representation (IR)
of the JSON filter.
This modular implementation allows for extendability in regards to file

> formats. All that is needed is a compatible front-end.

Reviewer: ```suggestion
formats. All that is needed is a compatible frontend.
```

Actors 1 and 2 in the diagram below illustrate the steps required for the JSON
and Rust filters to be compiled into BPF, respectively.

> ![seccompiler architecture](./docs/img/seccompiler_arch.jpg)

Reviewer 2: I think this needs to be updated to have input and output represented on the arrows. Let's chat offline.

### Filter definition

Let us take a closer look at what a filter is composed of, and how it is
defined:

> The smallest unit of the filter is the `SeccompCondition`, which is a

Reviewer 2: I think it might help to also show an example of each of these after their definition.

rule is a collection of **and-bound** conditions for a system call.

Finally, at the top level, there’s the `SeccompFilter`. The filter can be
viewed as a collection of syscall-associated rules, with a predefined on-match
action and a default action that is returned if none of the rules match.

In a filter, each system call number maps to a vector of **or-bound** rules.
In order for the filter to match, it is enough that one rule associated to the
system call matches. A system call may also map to an empty rule vector, which
means that the system call will match, regardless of the actual arguments.

The following diagram models a simple filter, that only allows `accept4`,
`fcntl(any, F_SETFD, FD_CLOEXEC, ..)` and `fcntl(any, F_GETFD, ...)`.
For any other system calls, the process will be killed.

![filter diagram](./docs/img/seccomp_filter.png)

As specified earlier, there are two ways of expressing the filters:

1. JSON (documented in [json_format.md](docs/json_format.md));
2. Rust code (documented by the library).

See below examples of both representation methods, for a filter equivalent to
the diagram above:

#### Example JSON filter

> ```json

Reviewer 2: It's generally better to say a few words about the example before it to prepare the readers. For example, here we can write something like:

Below you can find a filter example where the default action is to kill the process when a syscall that is not one of accept4 or fnctl is executed.

Reviewer 2: Nvm, I completely missed the above phrase that says exactly that :laughing:

{
    "main_thread": {
        "default_action": "kill_process",

>         "filter_action": "allow",

Reviewer 2: I am having difficulties understanding what is the difference between `default_action` and `filter_action`.

            },
            {
                "syscall": "fcntl",
                "args": [
                    {
                        "index": 1,
                        "type": "dword",
                        "op": "eq",
                        "val": 2,
                        "comment": "F_SETFD"
                    },
                    {
                        "index": 2,
                        "type": "dword",
                        "op": "eq",
                        "val": 1,
                        "comment": "FD_CLOEXEC"
                    }
                ]
            },
            {
                "syscall": "fcntl",
                "args": [
                    {
                        "index": 1,
                        "type": "dword",
                        "op": "eq",
                        "val": 1,
                        "comment": "F_GETFD"
                    }
                ]
            }
        ]
    }
}
```

Note that the JSON files need to specify a name for each filter. In the
example above, there is only one: `main_thread`, other programs may be using
multiple filters.

#### Example Rust-based filter

```rust

> SeccompFilter::new(

Reviewer 2: Since we need to add comments about what parameter represents, maybe this is a good indicator that we should make the initialization easier.

When parameters have different types making mistakes in the initialization is hard, but in this particular case, both filter_action and default_action have the same type (SeccompAction). This makes it prone to errors as you can reverse the two without realizing.

The easiest way out is to use the builder pattern, but there might be other options worth exploring.

Reviewer 2: I think we can handle this in the code review, and update the documentation afterwards.

>     // default_action

Reviewer 2: Are there non default actions supported? If not, maybe we can name this `action` instead of `default_action`.

location (e.g. using `env::var("OUT_DIR")`) the compiled filters, that have
been serialized to a binary format
(e.g. [bincode](https://github.com/bincode-org/bincode)).
They can then be ingested by the application using `include_bytes!` and
deserialized before getting installed.
This build-time option can be used to shave off the filter compilation time
from the app startup time, if using a low-overhead binary format.

Regardless of the compilation moment, the process is the same:

For **JSON filters**, the compilation to loadable BPF is performed using the
`compile_from_json()` function:

```rust
let filters: BpfMap = seccompiler::compile_from_json(
    File::open("/path/to/json")?, // Accepts generic Read objects.
    seccompiler::TargetArch::x86_64,
)?;
```

`BpfMap` is another type exposed by the library, which maps thread
categories to BPF programs.

```rust
pub type BpfMap = HashMap<String, Arc<BpfProgram>>;
```

For **Rust filters**, it’s enough to perform a `try_into()` cast, from a
`SeccompFilter`, to a `BpfProgram`, like so:

```rust
let seccomp_filter = SeccompFilter::new(
    rule_map,
    SeccompAction::Trap,
    seccompiler::TargetArch::x86_64
)?;

> let bpf_prog: BpfProgram = seccomp_filter.try_into()?;

Reviewer 2: Is this the format expected by the kernel?

Author: yes, it's a Vec<sock_filter> and the kernel expects a `sock_filter` pointer
